libsodium-sys/build.rs

```diff
@@ -42,55 +94,406 @@ fn main() {
+        let opt_install_path = if entry_path.starts_with("include") {
+            let is_dir = (unwrap!(entry.unix_mode()) & S_IFDIR as u32) != 0;
+            if is_dir {
+                let _ = fs::create_dir(&Path::new(&install_dir).join(entry_path));
+                None
+            } else {
+                Some(Path::new(&install_dir).join(entry_path))
+            }
+        } else if entry_path == unpacked_lib {
+            Some(lib_install_dir.join("libsodium.lib"))
+        } else {
+            None
+        };
+        if let Some(full_install_path) = opt_install_path {
+            let mut buffer = Vec::with_capacity(entry.size() as usize);
+            assert_eq!(entry.size(), unwrap!(entry.read_to_end(&mut buffer)) as u64);
+            let mut file = unwrap!(File::create(&full_install_path));
+            unwrap!(file.write_all(&buffer));
+        }
+    }
+
+    println!(
+        "cargo:rustc-link-search=native={}",
+        lib_install_dir.display()
+    );
+
+    Some(format!("{}/include", install_dir))
+}
+
+#[cfg(all(windows, not(target_env = "msvc")))]
+fn build_libsodium() -> Option<String> {
+    use flate2::read::GzDecoder;
+    use tar::Archive;
+
+    // Download gz tarball
+    let install_dir = get_install_dir();
+    let lib_install_dir = Path::new(&install_dir).join("lib");
+    unwrap!(fs::create_dir_all(&lib_install_dir));
+    let url = format!("{}libsodium-{}-mingw.tar.gz", DOWNLOAD_BASE_URL, VERSION);
+    let compressed_file = download(&url, SHA256);
+
+    // Unpack the tarball
+    let gz_decoder = GzDecoder::new(compressed_file);
+    let mut archive = Archive::new(gz_decoder);
+
+    // Extract just the appropriate version of libsodium.a and headers to the install path
+    let arch_path = if cfg!(target_pointer_width = "32") {
+        Path::new("libsodium-win32")
+    } else if cfg!(target_pointer_width = "64") {
+        Path::new("libsodium-win64")
+    } else {
+        panic!("target_pointer_width not 32 or 64")
+    };
+
+    let unpacked_include = arch_path.join("include");
+    let unpacked_lib = arch_path.join("lib\\libsodium.a");
+    let entries = unwrap!(archive.entries());
+    for entry_result in entries {
+        let mut entry = unwrap!(entry_result);
+        let entry_path = unwrap!(entry.path()).to_path_buf();
+        let full_install_path = if entry_path.starts_with(&unpacked_include) {
+            let include_file = unwrap!(entry_path.strip_prefix(arch_path));
+            Path::new(&install_dir).join(include_file)
+        } else if entry_path == unpacked_lib {
+            lib_install_dir.join("libsodium.a")
+        } else {
+            continue;
+        };
+        unwrap!(entry.unpack(full_install_path));
+    }
+
+    println!(
+        "cargo:rustc-link-search=native={}",
+        lib_install_dir.display()
+    );
+
+    Some(format!("{}/include", install_dir))
+}
+
+#[cfg(not(windows))]
+fn build_libsodium() -> Option<String> {
```

Comment:
It seems `build_libsodium` can't return `None`? If so then maybe return just `String`?

Reply:
updated :+1: